added oracledb support

petronny opened the pull request:
cx_Oracle has a major new release under a new name and homepage [python-oracledb](https://oracle.github.io/python-oracledb/).

The python-oracledb driver is a [Python programming language](https://www.python.org/) extension module allowing Python programs to connect to [Oracle Database](https://www.oracle.com/database/). Python-oracledb is the new name for Oracle's popular cx_Oracle driver.

Tested with an Oracle Autonomous Database.

felixxm commented on django/db/backends/oracle/base.py:
```diff
@@ -49,9 +49,12 @@ def _setup_environment(environ):
 except ImportError as e:
-    raise ImproperlyConfigured("Error loading cx_Oracle module: %s" % e)
+    try:
+        import cx_Oracle as Database
```
This should raise a deprecation warning, see [comment](https://code.djangoproject.com/ticket/33817#comment:1) and [docs about deprecating a feature](https://docs.djangoproject.com/en/4.0/internals/contributing/writing-code/submitting-patches/#deprecating-a-feature).

felixxm replied:
Deprecation should be documented, please see guide that I mentioned :point_up:

felixxm replied:
Tests of deprecation warnings are missing.

petronny replied:
Hi, I've updated the documents. But I don't know how to invoke the when deprecation warnings adding the tests since we are just importing libraries. Do you have any suggestion?

petronny replied:
Also, do I need to add this warning in the files like `django/contrib/gis/db/backends/oracle/adapter.py`, `django/contrib/gis/db/backends/oracle/introspection.py` and `django/db/backends/oracle/introspection.py`?
Or only add it in `django/db/backends/oracle/base.py`?

felixxm replied:
> Also, do I need to add this warning in the files like `django/contrib/gis/db/backends/oracle/adapter.py`, `django/contrib/gis/db/backends/oracle/introspection.py` and `django/db/backends/oracle/introspection.py`? Or only add it in `django/db/backends/oracle/base.py`?

Yes, we should have warnings in both backends `django.db.backends.oracle` and `django.contrib.gis.db.backends.oracle`. I'd prefer to have a single point (`base.py`) that will raise a warning
```python
try:
    import oracledb
except ImportError:
    import cx_Oracle as oracledb
    warnings.warn(
        "cx_Oracle is deprecated. Please use oracledb instead.",
        RemovedInDjango51Warning,
    )
```
In other imports, switching between libraries should be sufficient.

felixxm replied:
This is still untested.

felixxm commented on docs/ref/databases.txt:
```diff
@@ -857,11 +857,10 @@ To enable the JSON1 extension you can follow the instruction on
 
-Django supports `Oracle Database Server`_ versions 19c and higher. Version 7.0
-or higher of the `cx_Oracle`_ Python driver is required.
+Django supports `Oracle Database Server`_ versions 19c and higher. The `oracledb`_ Python driver is required.
```
Wrap docs at 79 chars.

felixxm commented:
@petronny We should change `tests/requirements/oracle.txt`.

petronny commented on django/contrib/gis/db/backends/oracle/adapter.py:
```diff
@@ -1,4 +1,14 @@
+    warnings.warn(
+        "cx_Oracle is deprecated. Please use oracledb instead.",
+        RemovedInDjango51Warning,
+    )
```
Hi, do I need to add this warning in files except than `django/db/backends/oracle/base.py`, such as `django/contrib/gis/db/backends/oracle/adapter.py`, `django/contrib/gis/db/backends/oracle/introspection.py` and `django/db/backends/oracle/introspection.py`?

felixxm commented:
buildbot, test on oracle.

felixxm commented on django/db/backends/oracle/introspection.py:
```diff
@@ -17,39 +20,23 @@ class DatabaseIntrospection(BaseDatabaseIntrospection):
+            oracledb.DB_TYPE_NUMBER: "DecimalField",
+            oracledb.DB_TYPE_TIMESTAMP: "DateTimeField",
+            oracledb.DB_TYPE_VARCHAR: "CharField",
+        }
```
You cannot remove the `cx_Oracle` < 8 branch, it's deprecated but still supported.

petronny replied:
Oops, forgot that. Just fixed.

petronny commented on django/db/backends/oracle/introspection.py:
```diff
@@ -17,39 +20,42 @@ class DatabaseIntrospection(BaseDatabaseIntrospection):
+        # Legacy support for cx_Oracle 7.0
+        if self.connection.oracledb_version >= (
+            7,
+        ) and self.connection.oracledb_version < (8,):
```
Suggested formatting from black.

felixxm replied:
I'd add `HAS_CX_ORACLE` flag to use it here:
```python
try:
    import oracledb
except ImportError:
    import cx_Oracle as oracledb
    HAS_CX_ORACLE = True
else:
    HAS_CX_ORACLE = False

...

   @cached_property
    def data_types_reverse(self):
        # RemovedInDjango51Warning.
        if HAS_CX_ORACLE and self.connection.cx_oracle_version < (8,):
```

felixxm commented:
buildbot, test on oracle.

felixxm commented:
@petronny Please push fixes in separate commits. It's hard to review after force-push.

petronny commented:
> Please push fixes in separate commits.

OK.

The last change only replaces `oracledb.OBJECT` with `oracle.DB_TYPE_OBJECT`. These two are same but the first one is deprecated.

This won't fix the failed test in oragis. I've created an issue https://github.com/oracle/python-oracledb/issues/43 for it on oracledb.

petronny commented:
https://github.com/oracle/python-oracledb/issues/43#issuecomment-1206453289

According the reply of the oracledb group,
this type is not currently supported but may be supported in the future.

For now, we have to enable the thick mode of oracledb which needs the Oracle Client libraries.

Are these libraries available in oragis19 tests?

cjbj commented:
> For now, we have to enable the thick mode of oracledb to pass this test, which needs the Oracle Client libraries. Are these libraries available in the oragis19 test?

I'm guessing that since cx_Oracle was tested, then the libraries are available.

Overall, I recommend skipping this specific test when python-oracledb is in Thin mode (until the future time that the feature is supported in Thin mode).  You could do two test runs: the first with Thin mode and the second with Thick mode.

petronny commented:
I've updated the code to enable this test only for the thick mode of oracledb and cx_Oracle.

But I have no idea how to setup a test for the thick mode.

suraj-ora-2020 commented:
> But I have no idea how to setup a test for the thick mode.
Since you were able to run cx_Oracle so i assume you have the oracle instant client installed. So to switch from thin mode to thick mode in pyoracledb you need to call oracledb.init_oracle_client(lib_dir="path_to_instant_client"). You just need to call it once. You can call it either in the application or you can call after import oracledb in the following file django/db/backends/oracle/base.py.

suraj-ora-2020 commented:
> I've updated the code to enable this test only for the thick mode of oracledb and cx_Oracle.
> 
> But I have no idea how to setup a test for the thick mode.

Since you were able to run cx_Oracle so i assume you have the oracle instant client installed. So to switch from thin mode to thick mode in pyoracledb you need to call oracledb.init_oracle_client(lib_dir="path_to_instant_client"). You just need to call it once. You can call it either in the application or you can call after import oracledb in the following file django/db/backends/oracle/base.py.

petronny commented:
> You can call it either in the application or you can call after import oracledb in the following file django/db/backends/oracle/base.py.

It will force the mode to be thick if called in base.py.
And I believe that most users should perfer the thin mode.
So I perfer to call it just in the application.

But now we need some new jenkins workflows (or pipelines, idk) to test both the thin the thick mode, which are:

* oracle19, for thin mode and the basic fields.
* oragis19, for thin mode and the geometry fields.
* oracle19-thick, for thick mode and the basic fields.
* oragis19-thick , for thick mode and the geometry fields.

For the first two we can just use the existing oracle19 and oragis19 tests.
I have no idea how to add the last two for jenkins and enable the thick mode in them.

cjbj commented:
@felixxm are you able to help with @petronny's question about enabling extra tests?

felixxm commented:
> @felixxm are you able to help with @petronny's question about enabling extra tests?

IMO, we don't want to maintain multiple workflows for Oracle. We can decide to always use `thick` mode as `thin` mode is limited :thinking:. I should have some time to check this carefully in the coming weeks.

petronny commented:
Well I wish there could be an option to still use to the thin mode if the thick mode is set as the default since I don't need the geometry field.

suraj-ora-2020 commented:
> 

@felixxm I think using default as thick would not server the benefit for which pyoracledb has been developed.In thin mode we dont require instant client library dependencies, its light weight as well faster. We are continuously working on expanding thin mode functionality. We can give option to users to choose between thin and thick mode using some parameters in the settings.py.

suraj-ora-2020 commented:
> 



> > @felixxm are you able to help with @petronny's question about enabling extra tests?
> 
> IMO, we don't want to maintain multiple workflows for Oracle. We can decide to always use `thick` mode as `thin` mode is limited 🤔. I should have some time to check this carefully in the coming weeks.

@felixxm I think using default as thick would not server the benefit for which pyoracledb has been developed.In thin mode we dont require instant client library dependencies, its light weight as well faster. We are continuously working on expanding thin mode functionality. We can give option to users to choose between thin and thick mode using some parameters in the settings.py.

cjbj commented:
To get some progress on this PR, I can settle for using Thick mode until Thin mode has the Object support needed for all tests.  Users (like us) will be able to keep testing Thin mode easily enough - particularly if this PR lands.

petronny commented:
@felixxm Hi, I have tried to enable the thick mode in `tests/dbshell/test_oracle.py` but I'm not sure if it will work. If it works then the PR is ready for review and testing on the oracle oracle database.

suraj-ora-2020 commented:
@petronny I had ran the test suite against your PR and have got some observations.
The stats with you PR is:
15964 tests
FAILED (failures=106, errors=41, skipped=1755, expected failures=6).

The stats with the current django 4.2dev version is:
15969 tests
FAILED (failures=24, errors=3, skipped=1755, expected failures=6)

As we can see some new failures are coming with the PR which were not present earlier. Most of the failures are exactly same and repetative.
AssertionError: 318 != 0 : Stream should be empty: actually contains '/Users/apple/oracle/django_mod1/django/django/db/backends/oracle/base.py:57

Test names : 
test_base_command_multiple_label (admin_scripts.tests.CommandTypes),
test_testrunner_option (test_runner.tests.CustomTestRunnerOptionsCmdlineTests), etc.

Maybe some of the new changes is causing this failures.

petronny commented:
@suraj-ora-2020 Thanks for the tests.

`django/db/backends/oracle/base.py:57` gives a warning when oracledb is in the thin mode.
However the test wants no warning at all.

I've already called `oracledb.init_oracle_client()` in `tests/dbshell/test_oracle.py:15`. But maybe `django/db/backends/oracle/base.py` is loaded earlier than that.

Could you have a try to just remove the warning at `django/db/backends/oracle/base.py:57` to see if the tests pass?
If so then I just need to figure out a new place to give the warning.

suraj-ora-2020 commented:
> @suraj-ora-2020 Thanks for the tests.
> 
> `django/db/backends/oracle/base.py:57` gives a warning when oracledb is in the thin mode. However the test wants no warning at all.
> 
> I've already called `oracledb.init_oracle_client()` in `tests/dbshell/test_oracle.py:15`. But maybe `django/db/backends/oracle/base.py` is loaded earlier than that.
> 
> Could you have a try to just remove the warning at `django/db/backends/oracle/base.py:57` to see if the tests pass? If pass then I just need to figure out a new place to give the warning.



> @suraj-ora-2020 Thanks for the tests.
> 
> `django/db/backends/oracle/base.py:57` gives a warning when oracledb is in the thin mode. However the test wants no warning at all.
> 
> I've already called `oracledb.init_oracle_client()` in `tests/dbshell/test_oracle.py:15`. But maybe `django/db/backends/oracle/base.py` is loaded earlier than that.
> 
> Could you have a try to just remove the warning at `django/db/backends/oracle/base.py:57` to see if the tests pass? If pass then I just need to figure out a new place to give the warning.

@petronny The failures were due to the warnings. I tried removing the warnings and the tests were passing without failure.I guess the place for warnings need to be modified.

suraj-ora-2020 commented:
> @suraj-ora-2020 Thanks for the tests.
> 
> `django/db/backends/oracle/base.py:57` gives a warning when oracledb is in the thin mode. However the test wants no warning at all.
> 
> I've already called `oracledb.init_oracle_client()` in `tests/dbshell/test_oracle.py:15`. But maybe `django/db/backends/oracle/base.py` is loaded earlier than that.
> 
> Could you have a try to just remove the warning at `django/db/backends/oracle/base.py:57` to see if the tests pass? If pass then I just need to figure out a new place to give the warning.

@petronny The failures were due to the warnings. I tried removing the warnings and the tests were passing without failure.I guess the place for warnings need to be modified.

petronny commented:
> I guess the place for warnings need to be modified.

@suraj-ora-2020 I've moved that warning into DatabaseWrapper. Would you like to give it a new round of tests?

suraj-ora-2020 commented:
@petronny I can still see those errors coming from the modified place.
AssertionError: 335 != 0 : Stream should be empty: actually contains '/Users/apple/oracle/django_mod1/django/django/db/backends/oracle/base.py:251

petronny commented:
@suraj-ora-2020 Well, I've moved it to `DatabaseWrapper.init_connection_state()` just before the cursor is created. Could you test if it works this time?

suraj-ora-2020 commented:
> @suraj-ora-2020 Well, I've moved it to `DatabaseWrapper.init_connection_state()` just before the cursor is created. Could you test if it works this time?

@petronny Yes now it is working fine.

petronny commented:
Great! Now let's wait for the review from @felixxm.

suraj-ora-2020 commented:
@felixxm do we have any update on review or merging of this PR.

jacobtylerwalls commented:
> do we have any update on review or merging of this PR.

If this is ready for re-review, the ticket should be updated to remove the "Needs ..." flags _needs documentation_, _needs tests_, _patch needs improvement_.

petronny commented:
@jacobtylerwalls I see. Just done now.

suraj-ora-2020 commented:
Hi @petronny. In django/db/backends/oracle/base.py file in line 275 can you please modify the  Database.driver_mode.thin_mode to Database.is_thin_mode. 
https://python-oracledb.readthedocs.io/en/latest/api_manual/module.html#oracledb.is_thin_mode

jlost commented:
Really looking forward to this feature, and big kudos to Oracle for the thin driver!

suraj-ora-2020 commented:
@felixxm After the latest commit the test suite results are as expected and everything look good. Can you please have a final review for this PR.

jacobtylerwalls commented on django/db/backends/oracle/base.py:
```diff
@@ -49,11 +51,20 @@ def _setup_environment(environ):
+
+        warnings.warn(
+            "cx_Oracle is deprecated. Please use oracledb instead.",
+            RemovedInDjango51Warning,
```
```suggestion
            RemovedInDjango51Warning,
            stacklevel=2,
```
This is Django's usual practice, see e.g. ticket-18127.

petronny commented:
@suraj-ora-2020 Hi, could you help to test the geometry field support under the thin mode?

In https://github.com/oracle/python-oracledb/commit/0acdfdb5ad1e0ec1d6f0890df165f0ed96b35fbf , they add the support of `DB_TYPE_OBJECT` to the thin mode.

We can temporay change line 15, `if oracledb.is_thin_mode():`, in `django/contrib/gis/db/backends/oracle/features.py` to `if False` to enable the geometry field introspection test.

tysonclugg commented:
> We can change line 15, `if oracledb.is_thin_mode():`, in `django/contrib/gis/db/backends/oracle/features.py` to `if False:` to temporary enable the geometry field introspection test for the thin mode.

Or add a condition checking `distutils.version.LooseVersion(oracledb.version) >= distutils.version.LooseVersion('1.2')`.

cjbj commented:
@suraj-ora-2020  PTAL

suraj-ora-2020 commented:
> @suraj-ora-2020 Hi, could you help to test the geometry field support under the thin mode?
> 
> In [oracle/python-oracledb@0acdfdb](https://github.com/oracle/python-oracledb/commit/0acdfdb5ad1e0ec1d6f0890df165f0ed96b35fbf) , they add the support of `DB_TYPE_OBJECT` to the thin mode.
> 
> We can change line 15, `if oracledb.is_thin_mode():`, in `django/contrib/gis/db/backends/oracle/features.py` to `if False:` to temporary enable the geometry field introspection test for the thin mode.

@petronny I tested the django test suite with latest oracledb dev version(1.2dev) and the test results are inline. For geometry field test i am running the gis_tests module but its getting aborted. Do we have other specific tests which are testing the geometry field.

cjbj commented:
@petronny you probably saw that python-oracledb 1.2 was released today with the DB Object support feature.

suraj-ora-2020 commented:
@petronny For running the geometry field test we need to have the flag gis_enabled and supports_raster as True in the
 django/django/db/backends/base/features.py file. Also currently geometry field tests are written for postgresql only which uses postgis. Look at this django/django/blob/main/tests/gis_tests/rasterapp/migrations/0001_setup_extensions.py.

suraj-ora-2020 commented:
@petronny The current test suite does not test oracle DB_TYPE_OBJECT and those tests can be added in a future PR. Our priority is to land this PR so that it would be easy to test things in future. @felixxm We can skip the DB_TYPE_OBJECT tests for oracle now and move forward with this PR. Could you please have a final review for this PR.

petronny commented:
I agree with skipping the DB_TYPE_OBJECT tests for oracle now. I'll resolve the conflicts in this week.

felixxm commented:
buildbot, test on oracle.

felixxm commented on docs/internals/deprecation.txt:
```diff
@@ -35,10 +35,13 @@ details on these changes.
 * The ``map_width`` and ``map_height`` attributes of ``BaseGeometryWidget``
   will be removed.
 
+
```
Please revert unrelated change.

felixxm commented on django/db/backends/oracle/base.py:
```diff
@@ -259,6 +271,16 @@ def get_new_connection(self, conn_params):
     def init_connection_state(self):
         super().init_connection_state()
+
+        if Database.__name__ == "oracledb" and Database.is_thin_mode():
+            warnings.warn(
+                """
+                The support for the thin mode of oracledb is experimental. To switch
+                to the thick mode, please call oracledb.init_oracle_client() at the
+                beginning of your code.
+                """
+            )
```
Why this warning is needed? We shouldn't raise a warning when this mode is accepted by Django, if it's not acceptable then we should always call `init_oracle_client()` (also without a warning).

petronny replied:
Get it. Warning removed.

felixxm reviewed (commented):
@petronny Thanks :+1: I left comments. Also, we shouldn't use deprecated methods as `makedsn()` with `oracledb`.

petronny commented on tests/gis_tests/inspectapp/tests.py:
```diff
@@ -14,6 +14,7 @@
 
 
 class InspectDbTests(TestCase):
+    @skipUnlessDBFeature("supports_geometry_field_introspection")
```
Oops, I didn't notice that it's used in the code. I'll reverse these changes.

petronny commented on docs/ref/databases.txt:
```diff
@@ -861,11 +861,11 @@ To enable the JSON1 extension you can follow the instruction on
 
 .. _`Oracle Database Server`: https://www.oracle.com/
-.. _`cx_Oracle`: https://oracle.github.io/python-cx_Oracle/
+.. _`oracledb`: https://oracle.github.io/python-oracledb/
```
Added. But I'm sure if it's correct.

petronny commented:
Updated. And:

1. I still don't know how to test the warning for cx_oracle.
2. The new `oracle_any` codes need tests.
3. I plan to give the thin mode a try. If it's not working, I'll revert [this commit](https://github.com/django/django/pull/15841/commits/b91403f04aea788308119366501ca037f090533e) to stick with the thick mode in this PR.

HansAarneLiblik commented:
Has this had any thoughts / progress?

aqeelat commented on docs/releases/4.2.txt:
```diff
@@ -611,3 +611,7 @@ Miscellaneous
 * Passing positional arguments to ``Signer`` and ``TimestampSigner`` is
   deprecated in favor of keyword-only arguments.
+
+* Support for ``cx_Oracle`` is deprecated. Please use version 1.2.0 or higher of 
+  the ``oracledb`` instead.
```
```suggestion
* Support for ``cx_Oracle`` is deprecated. Please use version 1.2.0 or higher of
 ``oracledb`` instead.
```

cjbj commented:
Can we get progress on this PR?

felixxm commented:
> Can we get progress on this PR?

This PR must be rebased and retargeted to Django 5.0. When all issues and comments are solved, you can uncheck _"needs"_ flags on the ticket to put it back in the review queue.

cjbj commented:
> This PR must be rebased and retargeted to Django 5.0. When all issues and comments are solved, you can uncheck _"needs"_ flags on the ticket to put it back in the review queue.

@petronny are you able to do that?

petronny commented:
> This PR must be rebased and retargeted to Django 5.0.

Done. Feature will be added in 5.0 and deprecated in 6.0.

> When all issues and comments are solved, you can uncheck "needs" flags on the ticket to put it back in the review queue.

Hmmm, I always forgot to uncheck the flags before. Sorry.

@suraj-ora-2020 Could you show me a guide to set up a test enviroment for oracledb? I have an Oracle Autonomous Database but I need some help to run the tests.

suraj-ora-2020 commented:
@petronny I have created a doc for setup need to run the tests
https://docs.google.com/document/d/1dbJHQVkMznhR875kn1clTghVBdtlHHeJypFqLOq4puU/edit?usp=sharing.
Let me know if you face any issue

petronny commented:
> @petronny I have created a doc for setup need to run the tests https://docs.google.com/document/d/1dbJHQVkMznhR875kn1clTghVBdtlHHeJypFqLOq4puU/edit?usp=sharing. Let me know if you face any issue

@suraj-ora-2020 Thanks for your guide!

I just create a fresh autonomous database for the tests. And I have updated `YOUR{DSN,USERNAME,PASSWORD}` in the configuration. Specifically, `DATABASES['default']['USER']` and `DATABASES['other']['USER']` are set to `ADMIN` which is the default administrator account. I haven't created any other user yet.

But I still get `ORA-01031: insufficient privileges`, which is confusing me.

I see there are other settings about users like `DATABASES['default']['TEST']['USER']` and `DATABASES['other']['TEST']['USER']`.
Should I change them to another user and grant the priviledges?

Also, there is a user named `scott` mentioned in the guide. Should I also create a user named `scott` for the tests?

cjbj commented:
@suraj-ora-2020 can you provide the actual GRANT statement to make it easy for @petronny ?

petronny commented:
Seems that the always free Oracle Autonomous Database cannot provide the sufficient privileges to run the test.
I'm using a self-hosted oracle xe instead. The tests are running now.


> can you provide the actual GRANT statement to make it easy for @petronny ?

@cjbj Here is the script.

```sql
CREATE USER SCOTT IDENTIFIED BY tiger; 
GRANT CREATE TABLE TO SCOTT;
GRANT CREATE SEQUENCE TO SCOTT;
GRANT CREATE PROCEDURE TO SCOTT;
GRANT CREATE TRIGGER TO SCOTT;
GRANT CREATE USER TO SCOTT;
GRANT ALTER USER TO SCOTT;
GRANT DROP USER TO SCOTT;
GRANT CREATE TABLESPACE TO SCOTT;
GRANT DROP TABLESPACE TO SCOTT;
GRANT CREATE SESSION TO SCOTT WITH ADMIN OPTION;
GRANT CREATE TABLE TO SCOTT WITH ADMIN OPTION;
GRANT CREATE SEQUENCE TO SCOTT WITH ADMIN OPTION;
GRANT CREATE PROCEDURE TO SCOTT WITH ADMIN OPTION;
GRANT CREATE TRIGGER TO SCOTT WITH ADMIN OPTION;
GRANT EXECUTE ON SYS.DBMS_LOB TO SCOTT;
GRANT EXECUTE ON SYS.DBMS_RANDOM TO SCOTT;
```
or simply
```sql
CREATE USER SCOTT IDENTIFIED BY tiger; 
GRANT DBA TO SCOTT;
```

cjbj commented:
> @cjbj Here is the script.

I'll let @suraj-ora-2020 comment if anything can be improved or added.

I would suggest making sure to drop the new user at the end of the test

suraj-ora-2020 commented:
@cjbj  the test script destroyes the users automatically once the test has been completed.

felixxm commented on docs/ref/databases.txt:
```diff
@@ -916,11 +916,16 @@ To enable the JSON1 extension you can follow the instruction on
 
-Django supports `Oracle Database Server`_ versions 19c and higher. Version 7.0
-or higher of the `cx_Oracle`_ Python driver is required.
+Django supports `Oracle Database Server`_ versions 19c and higher. Version 1.1.0
```
`1.2.0`?
```suggestion
Django supports `Oracle Database Server`_ versions 19c and higher. Version 1.2.0
```

suraj-ora-2020 replied:
@felixxm Version 1.1.0 is ok. Version 1.1.0 or higher is required.

felixxm replied:
So this patch is consistent, we have different versions listed in different places.

suraj-ora-2020 replied:
I did not understand what do you mean by different versions listed in different places.

felixxm replied:
> I did not understand what do you mean by different versions listed in different places.

Please check this patch:
- `docs/releases/5.0.txt` -> version 1.2.0
- `tests/requirements/oracle.txt`  -> version 1.2.0
- `docs/ref/databases.txt` -> version  1.1.0

suraj-ora-2020 replied:
Thanks, understood now. Let us stick to 1.2.0 as DB_TYPE_OBJECT is supported from this version.

petronny commented:
About the test, I've locally tested on the following situations:

* python-oralcedb with thin mode
* python-oralcedb with thick mode
* cx_Oracle with the main branch

And they all shows `failures=23, errors=1, skipped=1245, expected failures=6` and I'm not sure if I need to fix them:
```
ERROR: test_key_quoted_string (model_fields.test_jsonfield.TestQuerying)
FAIL: test_deep_lookup_mixed (model_fields.test_jsonfield.TestQuerying)
FAIL: test_deep_lookup_objs (model_fields.test_jsonfield.TestQuerying)
FAIL: test_isnull_key_or_none (model_fields.test_jsonfield.TestQuerying)
FAIL: test_key_endswith (model_fields.test_jsonfield.TestQuerying)
FAIL: test_key_iendswith (model_fields.test_jsonfield.TestQuerying)
FAIL: test_key_iexact (model_fields.test_jsonfield.TestQuerying)
FAIL: test_key_in (model_fields.test_jsonfield.TestQuerying) (lookup='value__foo__in', value=['bar'])
FAIL: test_key_in (model_fields.test_jsonfield.TestQuerying) (lookup='value__foo__in', value=['bar', 'baz'])
FAIL: test_key_iregex (model_fields.test_jsonfield.TestQuerying)
FAIL: test_key_istartswith (model_fields.test_jsonfield.TestQuerying)
FAIL: test_key_regex (model_fields.test_jsonfield.TestQuerying)
FAIL: test_key_startswith (model_fields.test_jsonfield.TestQuerying)
FAIL: test_key_text_transform_char_lookup (model_fields.test_jsonfield.TestQuerying)
FAIL: test_lookup_exclude (model_fields.test_jsonfield.TestQuerying)
FAIL: test_lookup_exclude_nonexistent_key (model_fields.test_jsonfield.TestQuerying)
FAIL: test_nested_key_transform_on_subquery (model_fields.test_jsonfield.TestQuerying)
FAIL: test_none_key (model_fields.test_jsonfield.TestQuerying)
FAIL: test_none_key_and_exact_lookup (model_fields.test_jsonfield.TestQuerying)
FAIL: test_none_key_exclude (model_fields.test_jsonfield.TestQuerying)
FAIL: test_obj_subquery_lookup (model_fields.test_jsonfield.TestQuerying)
FAIL: test_ordering_grouping_by_key_transform (model_fields.test_jsonfield.TestQuerying)
FAIL: test_shallow_obj_lookup (model_fields.test_jsonfield.TestQuerying)
FAIL: test_slice_subquery_and_query (queries.tests.SubqueryTests)
```

felixxm commented:
buildbot, test on oracle.

felixxm commented:
> cx_Oracle with the main branch

If you have failures with `cx_Oracle` and Oracle 21c on the `main` branch than we should treat them as a separate issues and fix them in advance (CI use Oracle 19c). Please create a new ticket(s) for them in Trac and follow our [bug reporting guidelines](https://docs.djangoproject.com/en/dev/internals/contributing/bugs-and-features/#reporting-bugs). All bugfixes require [an accepted ticket](https://code.djangoproject.com/newticket).

petronny commented:
Hi, I've fixed the import issues in the [previous oracle gis tests](https://djangoci.com/job/pull-requests-oracle/database=oragis19,label=oracle,python=python3.11/592/console). And the previous oracle tests raise a [segmentation fault](https://djangoci.com/job/pull-requests-oracle/database=oracle19,label=oracle,python=python3.11/592/console). I don't know how to fix it but maybe it will disappear in the thin mode. I have done the switch now so could you give a new test run on the thin mode?

And I've rebased the PR with the lastest main branch. However this introduces new javascripts tests which fail with `Could not find Chromium`. This PR does nothing about javascripts so should I need to care about this error?

felixxm commented:
buildbot, test on oracle.

petronny commented:
I've fixed the import again. But the segmentation fault also shows in the thin mode :(

suraj-ora-2020 commented:
@petronny the links above that you have added has expired or not working. Are these segmentation faults you are getting in running gis_tests against oracle using oracledb(thin). If yes, I also tried running those tests but instead of getting seg fault i am getting AttributeError: 'DatabaseOperations' object has no attribute 'geo_db_type'.

suraj-ora-2020 commented:
> > @petronny the links above that you have added has expired or not working. Are these segmentation faults you are getting in running gis_tests against oracle using oracledb(thin). If yes, I also tried running those tests but instead of getting seg fault i am getting AttributeError: 'DatabaseOperations' object has no attribute 'geo_db_type'.
> 
> The segmentation fault shows in both thin and thick mode in the basic Oracle test. I don't remember the results on the gis tests now. How about the thick mode on your machine?

By basic test what do you mean as earlier with the test suite you were getting failures=23, errors=1, skipped=1245, expected failures=6. So by which tests you are getting segmentation fault.

petronny commented:
> By basic test what do you mean as earlier with the test suite you were getting failures=23, errors=1, skipped=1245, expected failures=6. So by which tests you are getting segmentation fault.

Running `./runtests.py --settings=test_oracle` locally with a dockerized oracledb 21c shows `failures=23, errors=1, skipped=1245, expected failures=6`.
The official test workflow `pull-requests-oracle/database=oracle19,label=oracle,python=3.11` gives ths segmentation fault.

@felixxm Could you trigger a new test on the oracle database? The previous log is disappeared so I'll save a copy of it this time.

matwrob commented:
Hi All, Is there any timeline for that implementation?

petronny commented:
I'm still trying to reproduce the segmentation fault.

I'm downloading this file: Oracle Database 19c Enterprise Edition 19.3.0.0.0 - Long Term Release for Linux x86-64 (V982063-01.zip) from [Oracle Software Delivery Cloud](https://edelivery.oracle.com/osdc/faces/SoftwareDelivery#).

suraj-ora-2020 commented:
@petronny I tested the PR  as well with main branch using oracle 19c and python 3.11 . I am not getting segmentation fault.
And I am seeing no errors as well which were coming earlier.
```Ran 16275 tests in 5361.270s
OK (skipped=1703, expected failures=6)```

suraj-ora-2020 commented:
@petronny The log file you posted for the seg fault contains only python stack trace. Is it possible to get the C stack trace. Can we have this configuration locally. If yes can you please mention the configuration the CI is using so that i can replicate that.

petronny commented:
@suraj-ora-2020 Great!
@felixxm Is it possible to configure the CI to also use this dockerized version of oracledb?

suraj-ora-2020 commented:
@petronny @felixxm The crash is caused with oracledb thin or thick. Is it possible to get the C stack trace. Or do we have some other ways to get some progress for this MR?

suraj-ora-2020 commented:
@petronny This crash is caused with oracledb thin or thick? Is it possible to get the C stack trace for this crash. 
@felixxm what could be the other possible way to get some progress for this MR?

petronny commented:
> This crash is caused with oracledb thin or thick?

Both thin and the thick mode will cause the segmentation fault.

> Is it possible to get the C stack trace for this crash.

@felixxm Is it possible to increase the verbose level on the official CI?

charettes commented:
buildbot, test on oracle.

charettes commented:
Rebuilding on Oracle for folks to have access to the logs which have been purged since the last run.

petronny commented:
> Rebuilding on Oracle for folks to have access to the logs which have been purged since the last run.

Uploaded the log here to prevent being purged.
[consoleText-20230616.txt](https://github.com/django/django/files/11766737/consoleText-20230616.txt)

PS. Previous log: [consoleText-20230419.txt](https://github.com/django/django/files/11270841/consoleText.txt)

apollo13 commented:
> @suraj-ora-2020 The log is directly downloaded from the official CI.
> 
> I can succesfully set up an dockerized oracledb 19c now. The process is: ...

I am trying to reproduce this. Is there any way to get a oracledb 19c docker image without having to build it and whatnot? Like a literal `docker pull oracle:19c` would be great…

petronny commented:
Hmm, I don't have a personal docker account. But building the docker image is very fast since it directly uses the downloaded binaries without compilation.

suraj-ora-2020 commented:
@apollo13 Are you able to set up the oracle db 19c . If not you can follow the steps to have it locally.

1. Install docker- https://hub.docker.com/editions/community/docker-ce-desktop-mac
2. Install/Download Oracle Database 19c for Linux x86-64 (zip format)  https://www.oracle.com/database/technologies/oracle-database-software-downloads.html
3. Clone the docker project: git clone https://github.com/oracle/docker-images.git.
4. Put the installed binary of Oracle DB  into docker-images/OracleDatabase/SingleInstance/dockerfiles/19.3.0 folder where respective Dockerfile script exists. i.e, cp ~/Downloads/LINUX.X64_193000_db_home.zip  ~/ docker images/OracleDatabase/SingleInstance/dockerfiles/19.3.0/
5. Execute the build script: cd docker-images/OracleDatabase/SingleInstance/dockerfiles/
./buildDockerImage.sh -v 19.3.0 -e
6. docker run --name mydb19c -d -p 51521:1525 -p 55500:5500 -e ORACLE_PWD=mysecurepassword -e ORACLE_CHARACTERSET=AL32UTF8 oracle/database:19.3.0-ee
7. Check container is running: docker ps

Let me know if you need anything else.

petronny commented:
> 7. Check container is running: docker ps

Please also remember to wait for a few minutes until the PDB is up. You can check with `lsnrctl services`.

felixxm commented:
buildbot, test on oracle.

petronny commented:
This time the segmentation fault disappears in previous tests but shows in another test.
```
  test_handle_db_exception (view_tests.tests.test_debug.DebugViewQueriesAllowedTests.test_handle_db_exception) ... Fatal Python error: Segmentation fault
```
[consoleText.txt](https://github.com/django/django/files/12258122/consoleText.txt)

felixxm commented:
I noticed a behavior change in `oracledb`, that can cause issues. In `cx_Oracle`, sockets to connections that cannot be open (e.g. [password is invalid](https://github.com/django/django/blob/3434dbd39d373df7193ad006b970c09c1a909ea3/tests/backends/oracle/tests.py#L109-L117)) are closed automatically:
- `cx-Oracle==8.3.0`:

  ```
  $ python -Wall runtests.py backends.oracle.tests.TransactionalTests
  Testing against Django installed in '/django/django' with up to 8 processes
  Found 2 test(s).
  Creating test database for alias 'default'...
  Creating test user...
  System check identified no issues (0 silenced).
  ..
  ----------------------------------------------------------------------
  Ran 2 tests in 2.048s

  OK
  Destroying test database for alias 'default'...
  Destroying test user...
  Destroying test database tables...
  ```

- `oracledb==1.3.2`

  ```
  python runtests.py backends.oracle.tests.TransactionalTests
  Testing against Django installed in '/django/django' with up to 8 processes
  Found 2 test(s).
  Creating test database for alias 'default'...
  Creating test user...
  System check identified no issues (0 silenced).
  .Exception ignored in: <socket.socket fd=3, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('127.0.0.1', 38016), raddr=('127.0.0.1', 1521)>
  Traceback (most recent call last):
    File "/usr/local/lib/python3.10/traceback.py", line 236, in clear_frames
      tb.tb_frame.clear()
  ResourceWarning: unclosed <socket.socket fd=3, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('127.0.0.1', 38016), raddr=('127.0.0.1', 1521)>
  .
  ----------------------------------------------------------------------
  Ran 2 tests in 1.860s

  OK
  Destroying test database for alias 'default'...
  Destroying test user...
  Destroying test database tables...
  ```

anthony-tuininga commented:
> This time the segmentation fault disappears in previous tests but shows in another test.

Can you supply the stack trace for the segfault? If not, can you supply the details of your environment and how you are running the tests? So far we have not been able to replicate the segfault.

I did correct the resource warning but I sincerely doubt that is the source of the segfault. If it is, I suspect it is only indirectly responsible and the real culprit is elsewhere!

felixxm commented:
> > This time the segmentation fault disappears in previous tests but shows in another test.
> 
> Can you supply the stack trace for the segfault?

I run tests with `faulthandler` and `catchsegv`, check out the stack trace: https://gist.github.com/felixxm/43ddde7dbc7a6887bc3c5e79638e0652.

> I did correct the resource warning but I sincerely doubt that is the source of the segfault.

Thanks :+1:

anthony-tuininga commented:
That stack trace isn't terribly helpful, unfortunately. It suggests that the failure is occuring inside Django code that traverses the frame -- specifically where it is looking for sensitive variables, not something to do with python-oracledb directly. Maybe I am missing something, though? I don't see any stack trace that references python-oracledb C code at all. Can you turn off the sensitive variable setting so that that code doesn't run? Maybe that will prove more enlightening!

felixxm commented:
buildbot, test on oracle.

felixxm commented:
buildbot, test on oracle.

felixxm commented:
`view_tests.tests.test_debug.DebugViewQueriesAllowedTests.test_handle_db_exception` crashes when we introspect traceback frames after catching `DatabaseError`. Precisely when we try to get `f_back` (the next outer frame) and assign it to the `current_frame`:

https://github.com/django/django/blob/5aa4c0b67549f1d7ab649f53ac782d8de3e6a87c/django/views/debug.py#L274

On Jenkins `f_back` is missing after `ThinCursorImpl.execute` and it crashes when assigning `current_frame = current_frame.f_back`:
```
0: <frame at 0x7f1774ed3e20, file 'src/oracledb/impl/thin/cursor.pyx', line 138, code oracledb.thin_impl.ThinCursorImpl.execute>
1: None
```
on my local machine:
```
0: <frame at 0x7f234f53ad40, file 'src/oracledb/impl/thin/cursor.pyx', line 138, code oracledb.thin_impl.ThinCursorImpl.execute>
1: <frame at 0x7f234e91d300, file '/lib/python3.10/site-packages/oracledb/cursor.py', line 378, code execute>
2: <frame at 0x7f234ea2ebc0, file '/django/db/backends/oracle/base.py', line 553, code execute>
...
```

ngnpope commented:
So it looks like Python 3.11 changed stuff with the frame object:

https://docs.python.org/3/whatsnew/3.11.html#pyframeobject-3-11-hiding

It also looks like `python-oracledb` is using `cython` for which there is the following issue: https://github.com/cython/cython/issues/5222

anthony-tuininga commented:
Yes, `python-oracledb` is using `cython`. So does this mean something needs to change in Django to handle the changes to the frames? Is this specific to Python 3.11 then?

apollo13 commented:
Python code as Django uses it shall never segfault. If it does it is bug in Python or from the looks of it in this case cython / python-oracledb.

On Tue, Aug 8, 2023, at 19:06, Anthony Tuininga wrote:
> Yes, `python-oracledb` is using `cython`. So does this mean something 
> needs to change in Django to handle the changes to the frames? Is this 
> specific to Python 3.11 then?
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/django/django/pull/15841#issuecomment-1669995515>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAAT5C5VCPS6QP5DG6LTUWLXUJW25ANCNFSM53J2OIHA>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>

anthony-tuininga commented:
I tried running the test suite on my machine using a local database (running 21.3) and it raised a number of exceptions but didn't segfault. I don't know precisely how you are running the test suite and what you have in your settings file. If you are able to share that information I can try replicate the issue here. Currently I see the following results:

```
Ran 16529 tests in 2056.447s

FAILED (failures=44, errors=3, skipped=1255, expected failures=6)
```

This is the contents of my settings file:

```python
# This is an Oracle test settings file for use with the Django test suite.
#

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.oracle',
        'NAME': 'localhost/DB21PDB1',
        'USER': 'django',
        'PASSWORD': 'dev',
        'TEST': {
            'USER': 'default_django_test',
            'TBLSPACE': 'default_django_test_ts',
            'TBLSPACE_TMP': 'default_django_temp_ts'
        }
    },
    'other': {
        'ENGINE': 'django.db.backends.oracle',
        'NAME': 'localhost/DB21PDB1',
        'USER': 'django',
        'PASSWORD': 'dev',
        'TEST': {
            'USER': 'other_django_test',
            'TBLSPACE': 'other_django_test_ts',
            'TBLSPACE_TMP': 'other_django_temp_ts'
        }
    },
}

DEFAULT_AUTO_FIELD='django.db.models.BigAutoField'
USE_TZ=False
SECRET_KEY="my_secret"
```

If any of this is incorrect (for example, a lot of failures are about the "wrong" AutoField) then please advise. I'd like to get to the bottom of this issue if at all possible!

anthony-tuininga commented:
Ok. I tweaked the settings file to follow the suggested file from @petronny. I now get this:

```
Ran 16529 tests in 1947.899s

FAILED (failures=23, errors=1, skipped=1255, expected failures=6)
```

Running the test that causes the segfault runs just fine on my machine. This is on Linux x86_64, Python 3.11.4 with the python-oracledb 1.3.2 prebuilt binary that was released publicly. Adding print statements in the `get_traceback_frame_variables()` method shows that it is indeed being called without any difficulty. @felixxm, what platform does the CI use, what environment does it set, etc. Is there a configuration file somewhere? I'll see if I can find it myself but I am not familiar with Django build procedures.

ngnpope commented:
You can see the CI job here: [`django-oracle`](https://djangoci.com/job/django-oracle/)

The configuration of interest would be [`database=oracle19,label=oracle,python=python3.11`](https://djangoci.com/job/django-oracle/database=oracle19,label=oracle,python=python3.11/).

And looking at the [latest build](https://djangoci.com/job/django-oracle/database=oracle19,label=oracle,python=python3.11/lastBuild/), specifically the [full console output](https://djangoci.com/job/django-oracle/database=oracle19,label=oracle,python=python3.11/lastBuild/consoleFull), we can see that it uses:
- Ubuntu 20.04 LTS (Focal Fossa)
- CPython 3.11.1
- x86-64

It may be that this has been fixed in 3.11.2 - see [release notes](https://docs.python.org/release/3.11.4/whatsnew/changelog.html#python-3-11-2-final). This sounds suspiciously relevant:

> - [gh-99110](https://github.com/python/cpython/issues/99110): Initialize frame->previous in frameobject.c to fix a segmentation fault when accessing frames created by PyFrame_New().

If you can install Python 3.11.1 and reproduce and then see if 3.11.2 fixes the issue it'd be helpful to confirm.

anthony-tuininga commented:
I could download and install 3.11.1 but is there a reason why the CI job doesn't get upgraded to 3.11.4? I am unable to replicate with 3.11.4 and your releaase notes reference does indeed seem quite relevant. Cython is likely using PyFrame_New() to build its frames.

felixxm commented:
buildbot, test on oracle.

felixxm commented:
I bumped Python versions on CI nodes, we will see.

petronny commented:
All tests have passed with python 3.11.4. 🎉

1. Do we need to make a note stating that version 3.11.1 should not be used?
2. Do we need to also test the thick mode?

felixxm commented:
> Do we need to make a note stating that version 3.11.1 should not be used?

No, we only officially supports the latest version.

> Do we need to also test the thick mode?

TBH, we don't have the time and resources to extensively test for Oracle drivers.

petronny commented:
> TBH, we don't have the time and resources to extensively test for Oracle drivers.

Then do we need to note that we only support the thin mode?

cjbj commented:
> Then do we need to note that we only support the thin mode?

It should work fine in Thick mode (as it has been doing so for many years), so I don't think the word 'support' is useful to add.  You could say it was tested with Thin mode if you want to say anything.  If any Thick mode issues are logged we can address them.

felixxm commented on django/db/backends/oracle/introspection.py:
```diff
@@ -20,39 +19,40 @@ class DatabaseIntrospection(BaseDatabaseIntrospection):
     def data_types_reverse(self):
-        if self.connection.cx_oracle_version < (8,):
+        # Legacy support for cx_Oracle 7.0
+        if not is_oracledb and self.connection.oracledb_version < (8,):
```
We can bump the minimum supported versions of `cx_Oracle` in advance, #17153.

felixxm commented on tests/requirements/oracle.txt:
```diff
@@ -1 +1 @@
-cx_oracle >= 7.0
+oracledb >= 1.2.0
```
We can wait a bit, the feature freeze for Django 5.0 is on September, 18th.

cjbj replied:
I would set 1.3.2 as minimum supported.  This will give maximum testing time before Django 5.0.  The changes in 1.4 are around areas that won't impact Django users.

cjbj replied:
I'd argue about the actual impact, but suspect I won't win.

felixxm replied:
> I'd argue about the actual impact, but suspect I won't win.

Agreed, the actual impact is not huge. I will try to merge this in the next few days.

felixxm replied:
@cjbj Since I have your (Oracle) attention, have you ever considered [becoming](https://www.djangoproject.com/foundation/corporate-membership/join/) a DSF [corporate member](https://www.djangoproject.com/foundation/corporate-membership/)? We've been supporting the Oracle backend on our own for 18 years.

cjbj replied:
I'm happy to take that to management.  There might be some follow up questions.  Would you care to chat to me first?

felixxm replied:
> Would you care to chat to me first?

Sure, it's always good to talk, you can use my email, it's [public](https://github.com/felixxm).

cjbj replied:
@felixxm I'll set up a call for next week.

felixxm replied:
(Un)fortunately, I'm on vacation next week. We can try after August, 21st or you can contact directly with the DSF board. I'm sure they would love to chat (\cc @django/dsf-board).

cjbj replied:
@felixxm next week is busy for me too, so I can wait until you return.

felixxm replied:
> @felixxm next week is busy for me too, so I can wait until you return.

I'm available for a call if you still want to catch up (daily from 6 am UTC to 12 o'clock UTC). We can figure out another time-slot if it doesn't work for you.

cjbj replied:
@felixxm I've sent a calendar invitation to meet tomorrow.

suraj-ora-2020 replied:
@sarahboyce I only used this conversion as a reference for the version bump.

felixxm commented on django/db/backends/oracle/oracledb_any.py:
```diff
@@ -0,0 +1,21 @@
+        )
+        is_oracledb = False
+    except ImportError:
+        raise ImproperlyConfigured("Error loading oracledb module: %s" % e)
```
```suggestion
        raise ImproperlyConfigured("Error loading cx_Oracle module: %s" % e)
```

petronny replied:
The error `e` is raised when importing oracledb so the traceback is for "Error loading oracledb module".

"Error loading oracledb module" is better in most cases.  "Error loading cx_Oracle module" may be better when oracledb is not installed and cx_Oracle is installed but unloadable. But displaying "Error loading oracledb module" is also acceptable here since people will just install oracledb when seesing the message.

|  |oracledb is not installed|oracledb is installed but unloadable|
|---|----|----|
|__cx_Oracle is not installed__|"Error loading oracledb module" is better.|"Error loading oracledb module" is better.|
|__cx_Oracle is installed but unloadable__|"Error loading cx_Oracle module" may be better. But "Error loading oracledb module" is also acceptable to me.|"Error loading oracledb module" is better.|

felixxm commented:
Rebased.

felixxm commented:
@petronny Thanks :+1: Welcome aboard :boat:

felixxm commented:
buildbot, test on oracle.

cjbj commented:
Yay!  A big thanks to all.

krismohan commented:
Superb. Let us (@oracle) know if there are features that require attention.
